interfaces/builtin/iotedge.go

```diff
@@ -0,0 +1,55 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const iotedgeSummary = `allows Azure iotedge daemon and client`
+
+const iotedgeBaseDeclarationSlots = `
+  azure-iotedge:
+    allow-installation: false
+    deny-connection: true
+    deny-auto-connection: true
+`
+
+const iotedgeConnectedPlugAppArmor = `
+# Description: allow access to the iotedge daemon socket and
+# tasks to manage containers. This gives privileged access to the system
+# via iotedge's socket API to iotedged, which then talks to dockerd
+
+# Description: control and tracing of dockerd containers
+@{PROC}/[0-9]*/environ r,
+ptrace (read),
+capability sys_ptrace,
+capability dac_read_search,
+
+# Allow iotedge daemon sockets
+/{,var/}run/iotedge/mgmt.sock rw,
```

Comment:
last I looked at it, iotedge's socket was configurable and I didn't need to add this when I produced a snap of iotedge, did that change? or are there other applications which expect to talk to iotedge on this socket? if there are other applications that need to talk to these sockets I would lean towards iotedge exposing it's sockets over a content interface these days

Reply:
yes, you can configure, but then you diverge how things behave on classic and snap.
Good example, docker docker is configurable, yet we allow it under run
Also when you use custom location, one needs to add content interface to share the socket.
It's just not as clean....

Reply:
when we first worked on the docker snap, I don't think the socket location was configurable FWIW, but also I'm not sure that making snap packaging slightly more complicated by supplying config like this is sufficient reasoning to add sockets, I think the biggest reason we have used before when justifying allowing sockets like this is when other applications on classic expect to see the socket in the same location and those other applications are not snaps, docker is an example of this and I'm not convinced that there are non-snap applications that expect to see the iotedge sockets exposed at these directories (but perhaps there are). 

also arguably, using the content interface to share the socket is arguably a better solution for snaps as it allows you to do things like default-provider and more explicitly matches the concept of a snap providing the socket

we currently have the problem that too many interfaces (docker being one of them) allow access to the global thing, and the slots are implicitly provided by the core snap, but are now being snapped, and now a snap wants to provide a slot for that and we have the problem in snapd of deciding who gets to provide that slot. If these applications provided the resource/socket over a content interface, we would not have these problems at all and things would be much simpler